pkg/proxy/apis/kubeproxyconfig/validation/validation.go

```diff
@@ -140,17 +141,24 @@ func validateKubeProxyConntrackConfiguration(config kubeproxyconfig.KubeProxyCon
 	return allErrs
 }
 
-func validateProxyMode(mode kubeproxyconfig.ProxyMode, fldPath *field.Path) field.ErrorList {
+func validateProxyMode(config *kubeproxyconfig.KubeProxyConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	switch mode {
+
+	ipvsFeatureGate := fmt.Sprintf("%s=true", features.SupportIPVSProxyMode)
+	if config.Mode == kubeproxyconfig.ProxyModeIPVS && !strings.Contains(config.FeatureGates, ipvsFeatureGate) {
```

Comment:
It's not correct as well.

If people doesn't enable IPVS feature gateway, then IPVS is an unknown mode. Kube-proxy will firstly try iptables mode, then userspace mode...

Reply:
People can specify IPVS proxy mode, but kube-proxy will not start in IPVS mode if people forget enable IPVS feature gateway.

Reply:
xref: https://github.com/kubernetes/kubernetes/pull/56390

Hopefully can help you :)

Reply:
Thanks for clarifying ! But I disagree with your point here :)
As a user, when I want to use ipvs mode, what I expect is that the final running mode is what I configured. If I forget to configure something or mis-configure something, K8S should tell me what/where I'm wrong, rather than running in another mode silently...

Reply:
I think that's what the feature gateway works. For example, I have never heard that we should force k8s admin enable feature gateway "AllowExtTrafficLocalEndpoints"

when 

```
service.Spec.ExternalTrafficPolicy == api.ServiceExternalTrafficPolicyTypeLocal
```

Reply:
so I think that's why we should error out and tell users what should be configured when they want to use a feature.

I believe there will be users (and I'm one of them :smile: ) complaining on the current behavior in the codebase.

See some other examples in validation:
```go
if !utilfeature.DefaultFeatureGate.Enabled(features.HugePages) && source.EmptyDir.Medium == core.StorageMediumHugePages {
	allErrs = append(allErrs, field.Forbidden(fldPath.Child("emptyDir").Child("medium"), "HugePages medium is disabled by feature-gate for EmptyDir volumes"))
}
```

```go
} else if fsResourceIsEphemeralStorage(fs.Resource) && !utilfeature.DefaultFeatureGate.Enabled(features.LocalStorageCapacityIsolation) {
	allErrs = append(allErrs, field.Forbidden(fldPath, "Containers' ephemeral storage requests/limits disabled by feature-gate for Downward API"))
}
```

```go
if devices != nil && !utilfeature.DefaultFeatureGate.Enabled(features.BlockVolume) {
	allErrs = append(allErrs, field.Forbidden(fldPath.Child("volumeDevices"), "Container volumeDevices is disabled by feature-gate"))
	return allErrs
}
```

```go
if resourceName == core.ResourceEphemeralStorage && !utilfeature.DefaultFeatureGate.Enabled(features.LocalStorageCapacityIsolation) {
	allErrs = append(allErrs, field.Forbidden(limPath, "ResourceEphemeralStorage field disabled by feature-gate for ResourceRequirements"))
}
if helper.IsHugePageResourceName(resourceName) && !utilfeature.DefaultFeatureGate.Enabled(features.HugePages) {
	allErrs = append(allErrs, field.Forbidden(limPath, fmt.Sprintf("%s field disabled by feature-gate for ResourceRequirements", resourceName)))
}
```

Reply:
@thockin @bowei any thoughts here?

Reply:
Took another look, I'm leaning toward erring out if feature gate is disabled while validating the config. Silently falling back to other modes might seem mysterious to users. On the other hand, it makes sense to fall back to other modes when feature gate is enabled but ipvs is not usable (modules not loaded etc.).

For the externalTrafficPolicy example @m1093782566 posted, I think we also WON'T fall back when feature gate is disabled?

Reply:
> For the externalTrafficPolicy example @m1093782566 posted, I think we also WON'T fall back when feature gate is disabled?

Do you mean we should validate it as well?